Rewrite signTransaction from FIO app using experimental approach

danieloravec opened the pull request:
This version's functionality should be roughly equivalent to current FIO app's signTransaction. Test coverage is weak.

danieloravec committed:
Init hash

* Add integrity hash to transaction context
* Modify init part in signTransaction so that
  it only initializes transaction hash and
  integrity hash
* Add initHash instruction to js
* Add index2.ts with test, runnable by
  yarn run run-example-experimental

danieloravec committed:
End hash

* Set p1 parameter in SIGN_TX such that
  endHash is performed
* Add endHash to example

danieloravec committed:
Send constant to hash

* When p1 is set to correct constant in signTx,
  the sendDataToHashNoDisplay action is performed
* Add an example to ts

danieloravec committed:
Add tests for initHash, endHash, sendDataNoDisplay

danieloravec committed:
Add sendDataDisplay

* Add sendDataDisplay that gets header and body
  as parameters, displays them on ledger, adds
  header to integrity hash and body to tx hash
* Add sendDataDisplay to example

danieloravec committed:
Add test for sendDataDisplay

* Add allowed hash to list of allowed hashes
* Create sendDataDisplay.test.ts

danieloravec committed:
Add endHash to all tests

danieloravec committed:
Add header, body to sendDataNoDisplay

* Also add encoding to buffer and parse
  encoding from buffer, but encoding is not
  used in any way yet

danieloravec committed:
Parse buffers in sendDataDisplay, sendDataNoDisplay nicer

* Use structs to parse data

danieloravec committed:
Add encoding to sendData(Display/NoDisplay)

* Handle uint8_t and string differently
  in ts
* Handle uint8_t and string differently
  in signTransaction.c

danieloravec committed:
Use security policy correctly

* In sendDataNoDisplay
* In sendDataDisplay

danieloravec committed:
Add uint16, uint32, uint64 encodings

* Add those new encodings to ts as
  well as c
* To display uints on ledger
  and add them to tx hash,
  uint8, uint16 and uint32 are
  all saved as uint64 as well.
  For example, uint8 is displayed on the screen
  as uint64, while it goes as an uint8 to the tx hash

danieloravec committed:
Produce correct tx hash

* Set correct encodings and parse
  data correctly such that the appropriate
  data go to the tx hash
* Add ENCODING_DATETIME to ts only,
  it is not needed in c, as date is
  just a 4-byte uint
* Add ENCODING_HEX to handle data
  such as actor

danieloravec committed:
Merge sendDataDisplay/NoDisplay into sendData with parameter

* Last parameter of sendData is whether the data should
  be displayed
* Update example and tests accordingly

danieloravec committed:
Add test for signTransaction mainnet

danieloravec committed:
Add registers

* Add array of 8-byte registers
* Add initAction instruction
* Add endAction instruction
* Action sends the length of itself
  and index of a register first in an
  initAction instruction. The
  expected length of the action is
  written into the corresponding register.
  With each sendData that is part of a
  action description, a register index
  is sent. The corresponding register
  is decremented by the length of the
  data body. When an action description
  ends, the endAction instruction is called.
  It validates whether the value left
  in the register after all subtractions
  is 0.
* Non-action sendData instruction calls
  set registerIndex to NO_REGISTER
  constant
* The expected action length is
  added to the integrity hash

danieloravec committed:
Move example from index2 to index

danieloravec committed:
Add startFor instruction

* Not tested yet as endFor, startIteration and
  endIteration are still not available

danieloravec committed:
Add endFor function

* Add endFor to signTransaction
* Add startFor, endFor to example

danieloravec committed:
Add startIteration, endIteration

* Add them to C and js
* Create example with nested For with 1 iteration
* No tests

danieloravec committed:
Allow minK, maxK range in FOR

* Change the single allowed number of
  iterations to the inclusive range of
  [minNumIterations, maxNumIterations]
* This works with 0 too
* Both minNumIterations and maxNumIterations
  are added to the integrity hash

danieloravec committed:
Rename action to counted section

* Rename initAction instruction
  to startCountedSection
* Rename endAction instruction
  to endCountedSection
* This makes it more general
  and not FIO-specific

danieloravec committed:
Make counted section nesting possible

* Change START_COUNTED_SECTION,
  END_COUNTED_SECTION and SEND_DATA
  such that they allow for
  nested counted sections
* When an inner counted section ends,
  it increases the amount of
  data sent in the parent section
  by the amount of data it received.
  This way, data from inner counted sections
  are also counted towards all ancestor
  counted sections.
* Add example
* No tests

danieloravec committed:
Format code, remove redundant parts

* Use C formatter
* Wrap long lines
* Remove redundant commented
  and unused uncommented parts

danieloravec committed:
Demonstrate approach on multiple actions

* Multiple transfer fio actions can be
  in one tx

danieloravec committed:
Add interpreter

* Allowed txs are expressed using json
  inside src/tx-templates
* Interpreter that takes a template
  and data for this template was added
  into src/interpreter folder

danieloravec committed:
Fix tx signing, correct hash is produced now

* Fix encoding error in tx template json

relatko commented on ledger-app-fio/src/signTransaction.c:
```diff
@@ -78,714 +20,823 @@ void respondSuccessEmptyMsg()
 
-		ctx->expiration = u4be_read(wireData->expiration);
-		sha_256_append(&ctx->hashContext, (uint8_t *)&ctx->expiration, sizeof(ctx->expiration));
+		VALIDATE(ctx->sectionLevel < MAX_NESTED_COUNTED_SECTIONS, ERR_UNEXPECTED_INS);
```
The size of ctx->expectedSectionLength[ctx->sectionLevel] and similar variables is MAX_NESTED_COUNTED_SECTIONS+1. Why do we need these variables when ctx->sectionLevels is 0?

relatko commented on ledger-app-fio/src/signTransaction.c:
```diff
@@ -78,714 +20,823 @@ void respondSuccessEmptyMsg()
-		sha_256_append(&ctx->hashContext, (uint8_t *)&ctx->refBlockPrefix, sizeof(ctx->expiration));
+		ctx->sectionLevel++;
+		ctx->expectedSectionLength[ctx->sectionLevel] = parsedSectionLength;
+		ctx->currSectionLength[ctx->sectionLevel] = 0;
```
Merge these related variables into some struct.

relatko commented on ledger-app-fio/src/signTransaction.c:
```diff
@@ -78,714 +20,823 @@ void respondSuccessEmptyMsg()
-			CASE(ACTION_TYPE_TRNSFIOPUBKY, "Transfer FIO tokens");
-#	undef   CASE
-		default:
+	UI_STEP(HANDLE_SEND_DATA_DISPLAY)
```
This may be OK for now, but it needs some work. One can expect that this app will in the end have a lot of encodings and we need probably some simple common mechanism for them that separates encoding from other logic so when one wants to implement new encoding, he needs to touch the existing code as little as possible (ideally just editing some constants)

relatko commented on ledger-app-fio/src/signTransaction.c:
```diff
@@ -78,714 +20,823 @@ void respondSuccessEmptyMsg()
+	ctx->encoding = u1be_read(wireData1->encoding);
+	ctx->bodyBuf = (char *)wireData2->body;
+
+	if (ctx->encoding == ENCODING_STRING)
```
Again. We probably want to have here something separating encoding specific stuff.

relatko commented on ledger-app-fio/src/signTransaction.c:
```diff
@@ -78,714 +20,823 @@ void respondSuccessEmptyMsg()
+		memcpy(ctx->storageBuffer, ctx->bodyBuf, 64);
+		break;
+	case COMPARE_WITH_STORAGE:
+		VALIDATE(strncmp(ctx->storageBuffer, ctx->bodyBuf, wireData2->bodyLength[0]) == 0,
```
Both storageBuffer and bodyBuffer may contain 0 as far as I see. We probably want to save length and bytes.

relatko commented on ledger-app-fio/src/signTransaction.c:
```diff
@@ -78,714 +20,823 @@ void respondSuccessEmptyMsg()
+	switch (ctx->encoding)
+	{
+	case ENCODING_UINT8:
+		ctx->uint8Body = u1be_read(wireData2->body);
```
Do we need ctx->uint8Body? Why not just a local variable?

danieloravec committed:
Remove redundant context variables

* They were only needed in the original
  application, but are not used in this one

danieloravec committed:
Finalize ihash in each instruction

* Add new ctx variable prevHash for
  storing the integrity hash after the
  previous instruction
* When starting a new instruction an
  ihash is initialized, constants
  (ins code, ...) are added into it
  and then prevHash is added into it
* At the end of the instruction, ihash
  is finalized and saved into prevHash
* Fors have their own array of hashes,
  so prevHash is not used in starting
  and ending an iteration that much

danieloravec committed:
Change README to the experimental one

danieloravec committed:
Add test for tx signing for testnet

danieloravec committed:
Remove chainId from initHash

* Use sendData for sending it instead

danieloravec committed:
Update tests, add test for mainnet tx

danieloravec committed:
Writing hashing in js in progress

danieloravec committed:
Remove bodyLen from iHash in SEND_DATA

danieloravec committed:
Add storageInstruction to integrityHash

danieloravec committed:
Remove console logs, format code

danieloravec committed:
Fix readmes
